### DISCUSSION TITLE
TCP/IP communication example

### DISCUSSION THREAD

**Discussion by arniscg** [contiki-ng/contiki-ng#1788] [Ideas]
Hi.
I have noticed that there are not a single example using the TCP API.

For example using the following functions:
 - tcp_attach()
 - tcp_listen()
 - tcp_connect()
 
Also related to TCP, there are no examples for Protosockets library which looks quite interesting.
 
It's very hard to find any resource online about these APIs that are not more than 5-10 years old.
Like there is this [example](https://github.com/contiki-os/contiki/blob/master/examples/tcp-socket/tcp-server.c), but it looks like an old deprecated contiki repo.

Would it be possible to add some TCP client/server communication code examples?

**Comment by atiselsts** [comment]
The short answer is (in my subjective opinion) that we don't maintain examples which directly show TCP usage because there is no application demand for these examples. The original author of Contiki, Adam Dunkels had some great work on TCP in sensor networks in the 2000-ties, but TCP did not really catch on and isn't widely used in these days. UDP or non-IPv6 protocols are typically preferred.

We do have MQTT and websocket examples, as well as HTTP example in the `platform-specific` directory, which use TCP under the hood. TCP for these examples is more of a necessary evil (as its required by the protocol) than a feature.

**Reply by rahav** [reply]
Can you provide references to the mentioned non-IPv6 protocols that are available or used in Contiki?

**Comment by pjonsson** [comment]
@nvt has made a TCP socket test in #2025 that might be useful.
